File: algorithms/[1047]删除字符串中的所有相邻重复项/solution.ts
// topics = ["字符串"]

function removeDuplicates(S: string): string {
  const n = S.length;
  const arr = S.split('');

  for (let i = n - 2; i >= 0; i -= 1) {
    if (arr[i] == arr[i + 1]) {
      arr.splice(i, 2);
    }
  }

  return arr.join('');
}

File: algorithms/[1047]删除字符串中的所有相邻重复项/solution.py
# topics = ["字符串", "栈"]


from typing import List


class Solution:
    def removeDuplicates(self, S: str) -> str:
        """
        Stack
        time O(n), space O(n), n 为字符串长度
        """
        stack: List[str] = []

        for ch in S:
            if stack and stack[-1] == ch:
                stack.pop()
            else:
                stack.append(ch)

        return ''.join(stack)
